### PR TITLE
Adding default OPA policy

### LINKED ISSUE #490: OPA connector fails when no policies are defined
When no policies are defined a default should be used (e.g., allow) rather then erroring.

Currently this error shows up:
```
Could not select a read module for default/data-csv : rpc error: code = Unknown desc = error in GetOPAOperationDecision (i = 0): error in format of OPA evaluation (incorrect result map)
```

### PR DESCRIPTION
**Pull request by rohithdv**
Fixes #490.
Signed-off-by: rohithdv <rovallam@in.ibm.com>

### CODE REVIEW

**Comment by shlomitk1** [comment]
@rohithdv How does this work? This policy is loaded by default, so even if we have a user policy defined, a deny action will always be true, won't it?

**Comment by rohithdv** [comment]
> @rohithdv How does this work? This policy is loaded by default, so even if we have a user policy defined, a deny action will always be true, won't it?

Yes true. The earlier fix was not correct. I have updated the fix. Please check the latest changes in this PR. Thanks. @shlomitk1

**Review by shlomitk1 — APPROVED** [review]

**Review by ronenkat — COMMENTED** [review]
Is there a way to have the decision not in the code?
please document the default decision.
was this tested end-to-end?

**Comment by roee88** [comment]
> Is there a way to have the decision not in the code?

I personally prefer the approach here. The code must not fail if there are no rules. Thus the default is allow but to change it to deny by default you add a policy that denies if no other rules are triggered (currently not straight forward but with the new opa policies design it is).

 a policy to deny if there are no other rules triggered to make it deny by default.

**Comment by ronenkat** [comment]
> > Is there a way to have the decision not in the code?
> 
> I personally prefer the approach here. The code must not fail if there are no rules. Thus the default is allow but to change it to deny by default you add a policy that denies if no other rules are triggered (currently not straight forward but with the new opa policies design it is).

I would be more comfortable if the default would come from a file, either part of the git repo or a configmap, rather than compiled in the code.

**Comment by rohithdv** [comment]
> > > Is there a way to have the decision not in the code?
> > 
> > 
> > I personally prefer the approach here. The code must not fail if there are no rules. Thus the default is allow but to change it to deny by default you add a policy that denies if no other rules are triggered (currently not straight forward but with the new opa policies design it is).
> 
> I would be more comfortable if the default would come from a file, either part of the git repo or a configmap, rather than compiled in the code.

To enable default policy outside the opa connector (i.e., from a rego file), I was just experimenting with the below approach and I think the following solution may work. 

(1) Load the default_policy.rego (attached below) along with opa during opa installation phase. This is the default policy file which enables any default behaviour (say deny by default or allow by default) by using a "helper" variable called "enfaction"( short for enforcement action). The default policy file also takes care of any precedence conditions that we may define (for eg: deny > transform > allow). This is also shown as an example in the file. 

[defaultpolicy.rego.txt](https://github.com/IBM/the-mesh-for-data/files/6286280/defaultpolicy.rego.txt)

(2) Change the opa connector code (one time change only) to query the "enfaction" target (which is specified in the default_policy.rego) and process the opa response accordingly. 

(3) The users may load any further policies at their will (for example the policies given in user-created-policy-1 or user-created-policy-2 folders or any similar folder) and the corresponding policies will be enforced.  
 
Please let me know your feedback if this approach is ok. Thanks. @ronenkat @roee88.

**Comment by roee88** [comment]
@rohithdv your technical approach is similar to what I described (and did in the katalog PoC). But @ronenkat I'm sorry this is not a discussion for PR. I suggest to move along with the current *fix* (it's a bug being fixed) and move this discussion to the OPA+Taxonomy discussions. We need a fix before the release and the current one is legitimate IMO.

File: connectors/opa/lib/evaluate_opa.go [hunk 1]
```diff
@@ -75,5 +89,15 @@ func EvaluatePoliciesOnInput(inputMap map[string]interface{}, opaServerURL strin
 	log.Println("responsestring data")
 	log.Println(string(data))
 
-	return string(data), nil
+	currentData := string(data)
+	decisionid, flag := doesOpaHaveUserPoliciesLoaded(string(data))
```
Comment (ronenkat) [inline connectors/opa/lib/evaluate_opa.go:93]: the json parser expects []byte, why cast to string and back to []byte?

**Review by ronenkat — COMMENTED** [review]
there are fmt.printf and log.println in the code. Should we have both?
are the fmt statements for debugging? and they should be removed?

File: connectors/opa/lib/evaluate_opa.go (hunk 1 above)
Comment (rohithdv) [reply connectors/opa/lib/evaluate_opa.go]: Fixed. Please check. Thanks.

**Comment by rohithdv** [comment]
> there are fmt.printf and log.println in the code. Should we have both?
> are the fmt statements for debugging? and they should be removed?

Now using on log package for printing debug information. Removed fmt package for consistency. Thanks.

**Review by shlomitk1 — APPROVED** [review]
Verified.

**Pull request merged**